src/committee/orchestrator.rs

```diff
@@ -36,12 +40,235 @@ use super::node::{Round2Request, Round2Response};
+                if let Ok(resp) =
+                    serde_json::from_str::<bitcoincore_rpc::jsonrpc::Response>(&resp_str)
+                {
+                    if let Some(resp_data) = resp.result {
+                        resp_data.get().parse::<u64>().unwrap_or_default() == data
+                    } else {
+                        false
+                    }
+                } else {
+                    false
+                }
+            }
+        }
+    }
+
+    pub async fn keepalive_thread(state: Arc<RwLock<Self>>) {
+        debug!("Keepalive thread started");
+        loop {
+            // Sleep
+            sleep(Duration::from_secs(KEEPALIVE_WAIT_SECONDS)).await;
+            // Get array of members which are not *permanently* offline
+
+            let members_to_check = {
+                let r_lock = state.read().unwrap();
+                r_lock
+                    .key_to_addr
+                    .iter()
+                    .map(|(key, addr)| {
+                        let status = r_lock.status.get(key).unwrap().clone();
+                        (key.clone(), addr.clone(), status)
+                    })
+                    .filter(|(_, _, status)| match *status {
+                        MemberStatus::Online => true,
+                        MemberStatus::Disconnected((retry_time, _)) => {
+                            retry_time <= Self::get_current_time_secs()
+                        }
+                        _ => false,
+                    })
+                    .collect::<Vec<(frost_secp256k1_tr::Identifier, String, MemberStatus)>>()
+            };
+
+            // check alive for each one of them
+            let mut futures = Vec::with_capacity(members_to_check.len());
+            for member_data in members_to_check.iter() {
+                futures.push(Self::check_alive(member_data.1.clone()));
+            }
+
+            // resolve futures and update state
+            let keepalive_resp = join_all(futures).await;
+            for (idx, resp) in keepalive_resp.iter().enumerate() {
+                let (key, address, old_status) = &members_to_check[idx];
+                if *resp {
+                    if *old_status != MemberStatus::Online {
+                        let mut state_w = state.write().unwrap();
+                        let member_status = state_w.status.get_mut(key).unwrap();
+                        *member_status = MemberStatus::Online;
+                        info!("{address} is back online");
+                    } else {
+                        debug!("{address} is online");
+                    }
+                } else {
+                    let mut state_w = state.write().unwrap();
+                    let member_status = state_w.status.get_mut(key).unwrap();
+                    let last_retries = match old_status {
+                        MemberStatus::Disconnected((_, last_retry_number)) => *last_retry_number,
+                        _ => 0,
```

Comment:
note that it is possible that the node was marked as offline in the mean time (due to a failure to respond in round 2 of a signature), so here we might want to have 

```rust
match old_status {
                        MemberStatus::Disconnected((_, last_retry_number)) => *last_retry_number,
                        MemberStatus::Online => 0,
                        MemberStatus::Offline => continue,
}
```